Refactor GeoPoint and GeoShape with generics

craigtaverner opened the pull request:
In preparation for supporting CartesianPoint and CartesianShape
in aggregations, this PR adds a common interface between GeoPoint
and CartesianPoint, and then uses that to split out some key common
code that will be used in CartesianPoint and CartesianShape aggregations

craigtaverner committed:
Refactor GeoPoint and GeoShape with generics

In preparation for supporting CartesianPoint and CartesianShape
in aggregations, this PR adds a common interface between GeoPoint
and CartesianPoint, and then uses that to split out some key common
code that will be used in CartesianPoint and CartesianShape aggregations

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

iverase reviewed (commented):
I did a first pass on the points part and there is one thing I think we can improve. We are currently mixing geo classes with generic classes which makes the code confusing. 

For example `AbstractLeafGeoPointFieldData` is using internally MultiPointValues<GeoPoint> and I think it should be using `MultiGeoPointValues` instead. I think the trick  is not to use the generics extending ElasticPoint but using the generic extending  MultiPointValues<?>. like:

```
public abstract class LeafPointFieldData<T extends MultiPointValues<?>> implements LeafFieldData {

    /**
     * Return geo-point or point values.
     */
    public abstract T getPointValues();

    /**
     * Return the internal representation of geo_point or point doc values as a {@link SortedNumericDocValues}.
     * A point is encoded as a long that can be decoded by using
     * {@link org.elasticsearch.common.geo.GeoPoint#resetFromEncoded(long)}
     */
    public abstract SortedNumericDocValues getSortedNumericDocValues();

}
```

I think this will simplify quite a bit things and it will prevent some unnecessary castings.
This applies to other classes like IndexPointFieldData and AbstractPointIndexFieldData.

craigtaverner committed:
Simplify generics (by Ignacio)

Co-authored-by: Ignacio Vera <ivera@apache.org>

not-napoleon commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/plain/AbstractShapeIndexFieldData.java:
```diff
@@ -0,0 +1,73 @@
+import org.elasticsearch.xpack.spatial.index.fielddata.IndexShapeFieldData;
+import org.elasticsearch.xpack.spatial.index.fielddata.ShapeValues;
+
+public abstract class AbstractShapeIndexFieldData implements IndexShapeFieldData {
```
Javadoc please, ideally with some hints as to why and how one would build a concrete implementation of this.

not-napoleon commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/plain/AbstractShapeIndexFieldData.java:
```diff
@@ -0,0 +1,73 @@
+        XFieldComparatorSource.Nested nested,
+        boolean reverse
+    ) {
+        throw new IllegalArgumentException("can't sort on geo_shape field without using specific sorting feature, like geo_distance");
```
is it correct to specify `geo_shape` and `geo_distance` here?  might this abstraction have a non-geo specialization down the road?  I would prefer if it got the field name from the concrete instance, and didn't offer a suggestion for the sorting feature.

craigtaverner replied:
I've moved the exception into the extending class, where the specific names have meaning.

not-napoleon reviewed (changes requested):
This looks mostly okay to me, but the Writable ValuesSourceType is extremely concerning.  If it needs to be there, let's talk, and if it doesn't need to be there, I'll approve once it's dropped.

craigtaverner commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/stats/EnumCounters.java:
```diff
@@ -118,11 +118,9 @@ public String toString() {
         StringBuilder buf = new StringBuilder("[");
         boolean first = true;
         for (E e : enums) {
+            if (first == false) buf.append(", ");
```
Ha, ha! This was part of an earlier PR that Ignacio also pointed out was not relevant to that PR, so I pulled it into a separate, isolated PR at https://github.com/elastic/elasticsearch/pull/88505. But that has yet to be approved. Perhaps you could review it, and this trivial change can stop popping up in my PRs? 😸

craigtaverner replied:
At least it is removed from this PR now.

craigtaverner commented on server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSource.java:
```diff
@@ -744,9 +744,9 @@ public final MultiGeoPointValues geoPointValues(LeafReaderContext context) {
         public static class Fielddata extends GeoPoint {
 
-            protected final IndexGeoPointFieldData indexFieldData;
+            protected final IndexPointFieldData<org.elasticsearch.common.geo.GeoPoint> indexFieldData;
```
I believe this is resolved by your commit at https://github.com/elastic/elasticsearch/pull/89388/commits/26b11a6874897e56050f51b7855e2afa0733ba4b

craigtaverner commented on server/src/main/java/org/elasticsearch/common/geo/ElasticPoint.java:
```diff
@@ -0,0 +1,19 @@
+ * To facilitate maximizing the use of common code between GeoPoint and projected CRS
+ * we introduced this ElasticPoint as an interface of commonality.
+ */
+public interface ElasticPoint {
```
OK. Let's go for `SpatialPoint`. I think `BasePoint` sounds more like an abstract class than an interface, and `Point2d` sounds too much like the 2d classes in Lucene, which this is not related to at all.

craigtaverner commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/CellIdSource.java:
```diff
@@ -53,7 +53,7 @@ public final boolean isFloatingPoint() {
     public final SortedNumericDocValues longValues(LeafReaderContext ctx) {
         final MultiGeoPointValues multiGeoPointValues = valuesSource.geoPointValues(ctx);
-        final GeoPointValues values = org.elasticsearch.index.fielddata.FieldData.unwrapSingleton(multiGeoPointValues);
+        final GeoPointValues values = (GeoPointValues) org.elasticsearch.index.fielddata.FieldData.unwrapSingleton(multiGeoPointValues);
```
I believe this is resolved by your commit at https://github.com/elastic/elasticsearch/pull/89388/commits/26b11a6874897e56050f51b7855e2afa0733ba4b

craigtaverner commented on server/src/main/java/org/elasticsearch/common/geo/GeoUtils.java:
```diff
@@ -522,10 +522,10 @@ public static double planeDistance(double lat1, double lon1, double lat2, double
         final GeoPoint... fromPoints
     ) {
-        final GeoPointValues singleValues = FieldData.unwrapSingleton(geoPointValues);
+        final GeoPointValues singleValues = (GeoPointValues) FieldData.unwrapSingleton(geoPointValues);
```
I believe this is resolved by your commit at https://github.com/elastic/elasticsearch/pull/89388/commits/26b11a6874897e56050f51b7855e2afa0733ba4b

craigtaverner commented on server/src/main/java/org/elasticsearch/index/fielddata/plain/AbstractLeafGeoPointFieldData.java:
```diff
@@ -7,28 +7,35 @@
+    protected final ToScriptFieldFactory<MultiPointValues<GeoPoint>> toScriptFieldFactory;
 
-    public AbstractLeafGeoPointFieldData(ToScriptFieldFactory<MultiGeoPointValues> toScriptFieldFactory) {
+    public AbstractLeafGeoPointFieldData(ToScriptFieldFactory<MultiPointValues<GeoPoint>> toScriptFieldFactory) {
```
I believe this is resolved by your commit at https://github.com/elastic/elasticsearch/pull/89388/commits/26b11a6874897e56050f51b7855e2afa0733ba4b

craigtaverner commented on server/src/main/java/org/elasticsearch/common/geo/GeoBoundingBox.java:
```diff
@@ -10,107 +10,52 @@
-            builder.field(LAT_FIELD.getPreferredName(), topLeft.lat());
-            builder.field(LON_FIELD.getPreferredName(), topLeft.lon());
+            builder.field(LAT_FIELD.getPreferredName(), topLeft.getY());
+            builder.field(LON_FIELD.getPreferredName(), topLeft.getX());
```
I agree 100%. I want to move in a direction that isolates this kind of mapping to a single place. Right now there are many, many places that depend on secret knowledge of the mapping and the correct order (for example some methods take `(double y, double x)` and others take `(double x, double y)`. In the earlier Elastic code (and earlier Lucene code), `y, x` was the norm, but over time things have swung over to `x, y` being much more common. What I'm keen to do is iteratively minimize the differences. Some earlier PR's I did in the Spring already downplayed the `y, x` cases (in code and documentation), but that was a small step. This PR here now actually does improve things quite a bit (IMHO); but it also brings some things into the light, like the one you observe here. I think more can (and should) be done, but I do not want to change too much in one go. So this PR priorizes the changes required to enable the work on Cartesian, but does not do everything that could be done. This feels like a safer compromise.

craigtaverner replied:
And the creation of the `SpatialPoint` (neé `ElasticPoint`) interface leads to one option for where we could place this kind of knowledge. But I think we should think more about that before actually taking that step. There could be better options.

not-napoleon replied:
Sounds good.  I'm happy to keep this on the roadmap and address it in a future PR after more thought.

craigtaverner committed:
Refactor ElasticPoint to SpatialPoint

craigtaverner committed:
Rename ShapeValuesProvider to ShapeValuesSource

It extends ValuesSource, and is extended by GeoShapeValuesSource.
There is no reason for the suffix `Provider`.

craigtaverner commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/support/ShapeValuesProvider.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.io.IOException;
+
+public abstract class ShapeValuesProvider extends ValuesSource {
```
No idea. Seems wrong to me too. I've fixed it in 33cb378234087850893ef6d6815fe22d5971e3d7.

craigtaverner commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/GeoCentroid.java:
```diff
@@ -8,14 +8,7 @@
-
-    long count();
-}
+public interface GeoCentroid extends CentroidAggregation {}
```
This was because I wrote the CartesianCentroid stuff, and then Ignacio wanted the pure refactoring parts pulled out. The Cartesian work is in a separate PR on top of this work. If I move this change into that PR it will cause that PR to contain a lot of test code changes that are geo-specific (not cartesian). My goal is for this PR to contain only refactoring (moving Geo-specific code around) and the other PR to contain primarily Cartesian specific stuff (minimize Geo changes). So for the purpose of that split, this is a good change :-)

craigtaverner commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalCentroid.java:
```diff
@@ -0,0 +1,193 @@
+
+    protected abstract ElasticPoint centroidFromStream(StreamInput in) throws IOException;
+
+    protected abstract void centroidToStream(StreamOutput out) throws IOException;
```
These methods encapsulate the parts of the Geo vs. Cartesian that differ. In particular this is the `y, x` versus `x, y` we discussed in another comment. This move is a step towards trying to get that particular issue isolated. I think more can be done (eg. moving that knowledge into the `SpatialPoint` or some other class), but this step feels like a step in the right direction.

not-napoleon replied:
Yeah, I think they're fine, I just want them to have javadoc.

craigtaverner commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalGeoCentroid.java:
```diff
@@ -10,81 +10,72 @@
 
+    @Override
+    protected GeoPoint centroidFromStream(StreamInput in) throws IOException {
+        // TODO: Consider removing if we no longer need to support older versions than 7.17
```
OK. Makes sense. I've removed it.

craigtaverner commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/plain/AbstractShapeIndexFieldData.java:
```diff
@@ -0,0 +1,73 @@
+        throw new IllegalArgumentException("can't sort on geo_shape field without using specific sorting feature, like geo_distance");
+    }
+
+    public static class GeoBuilder implements IndexFieldData.Builder {
```
This builder is not needed, we can just use a lambda.

craigtaverner replied:
Deleted!

craigtaverner commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/plain/LatLonShapeIndexFieldData.java:
```diff
@@ -0,0 +1,78 @@
+        int bucketSize,
+        BucketedSort.ExtraData extra
+    ) {
+        throw new IllegalArgumentException("can't sort on geo_shape field without using specific sorting feature, like geo_distance");
```
I've done a minimal change, but not yet made use of `AggregationErrors`. I'll read up on that and learn about it.

craigtaverner committed:
Code review, mostly AbstractShapeIndexFieldData

craigtaverner committed:
Reverted trivial refactoring

craigtaverner committed:
Removed unused Writable interface implementation

craigtaverner commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/support/ShapeValuesSourceType.java:
```diff
@@ -0,0 +1,36 @@
+
+import java.io.IOException;
+
+public abstract class ShapeValuesSourceType implements Writeable, ValuesSourceType {
```
So, I thought a CI test of this would be a good idea. I pushed that in https://github.com/elastic/elasticsearch/pull/89388/commits/541e47f3069f7029693f76d3fee467e65f90d05a. Let's see how it goes and then make a decision.

craigtaverner commented on server/src/main/java/org/elasticsearch/index/fielddata/plain/LatLonPointIndexFieldData.java:
```diff
@@ -0,0 +1,88 @@
+        ValuesSourceType valuesSourceType,
+        ToScriptFieldFactory<MultiPointValues<GeoPoint>> toScriptFieldFactory
+    ) {
+        super(fieldName, valuesSourceType, toScriptFieldFactory);
```
I believe this is resolved by your commit at https://github.com/elastic/elasticsearch/commit/26b11a6874897e56050f51b7855e2afa0733ba4b

craigtaverner committed:
Further generics refinements

Based on Ignacio's work in 050df953dfed27c444c3821a6d939f7030dbcc16,
we fix the BoundingBox generics, and also add a little more specificity
to the previous generics (replace <?> with <? extends SpatialPoint>).

craigtaverner commented on server/src/main/java/org/elasticsearch/common/geo/GeoBoundingBox.java:
```diff
@@ -141,106 +86,36 @@ public void writeTo(StreamOutput out) throws IOException {
-        GeoPoint bottomRight = new GeoPoint(bottom, right);
-        return new GeoBoundingBox(topLeft, bottomRight);
+        GeoBoundsParser bounds = new GeoBoundsParser(parser);
+        return (GeoBoundingBox) bounds.parseBoundingBox();
```
OK. Done in https://github.com/elastic/elasticsearch/pull/89388/commits/c8b9ee3ed3b490a42dc7f64bb078ae083dbfafaf

craigtaverner commented on server/src/main/java/org/elasticsearch/common/geo/GeoBoundingBox.java:
```diff
@@ -10,107 +10,52 @@
-        return builder;
-    }
-
     public XContentBuilder toXContentFragment(XContentBuilder builder, boolean buildLatLonFields) throws IOException {
```
The value of buildLatLonFields is true almost always, and is also something geo-specific. We should get rid of it and have another method for the false case that only exists in the geo-implementation.

craigtaverner commented on server/src/main/java/org/elasticsearch/common/geo/BoundingBox.java:
```diff
@@ -0,0 +1,199 @@
+        return builder;
+    }
+
+    public abstract XContentBuilder toXContentFragment(XContentBuilder builder, boolean buildLatLonFields) throws IOException;
```
`buildLatLonField` sounds very geo-specific. We should be able to get rid of this for now.

craigtaverner commented on server/src/main/java/org/elasticsearch/common/geo/BoundingBox.java:
```diff
@@ -0,0 +1,199 @@
+    static final ParseField RIGHT_FIELD = new ParseField("right");
+    static final ParseField WKT_FIELD = new ParseField("wkt");
+    public static final ParseField BOUNDS_FIELD = new ParseField("bounds");
+    public static final ParseField LAT_FIELD = new ParseField("lat");
```
`LAT` and `LON` are geo-specific and should be in GeoBoundingBox

craigtaverner committed:
Removed some geo-specific code from BoundingBox

not-napoleon commented on server/src/main/java/org/elasticsearch/common/geo/BoundingBox.java:
```diff
@@ -0,0 +1,199 @@
+    static final ParseField LEFT_FIELD = new ParseField("left");
+    static final ParseField RIGHT_FIELD = new ParseField("right");
+    static final ParseField WKT_FIELD = new ParseField("wkt");
+    public static final ParseField BOUNDS_FIELD = new ParseField("bounds");
```
Yeah, I don't think we need to change it.  I just want a comment saying it's on purpose and why.

not-napoleon reviewed (approved):
This looks good.  As noted, there's follow up work, but this looks like a big improvement to me, so let's get it in.  Thanks for taking this on!
